File: docs/vsto/codesnippet/CSharp/trin_wordcontentcontrolreference/Group.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;

using Microsoft.VisualStudio.Tools.Applications.Runtime;
using Word = Microsoft.Office.Interop.Word;
using Office = Microsoft.Office.Core;

namespace Trin_ContentControlReference
{
    public partial class ThisDocument
    {
        // GroupContentControl constructor #1: create a GroupContentControl at the current selection.
        //<Snippet400>
        private Microsoft.Office.Tools.Word.GroupContentControl groupControl1;

        private void AddGroupControlAtSelection()
        {
            this.Paragraphs[1].Range.InsertParagraphBefore();
            Word.Range range1 = this.Paragraphs[1].Range;
            range1.Text = "You cannot edit or change the formatting of text " +
                "in this paragraph, because this paragraph is in a GroupContentControl.";
            range1.Select();

            groupControl1 = this.Controls.AddGroupContentControl("groupControl1");
        }
        //</Snippet400>

        // GroupContentControl constructor #2: create a GroupContentControl at a specified range.
        //<Snippet401>
        private Microsoft.Office.Tools.Word.GroupContentControl groupControl2;

        private void AddGroupControlAtRange()
        {
            this.Paragraphs[1].Range.InsertParagraphBefore();
            Word.Range range1 = this.Paragraphs[1].Range;
            range1.Text = "You cannot edit or change the formatting of text " +
                "in this paragraph, because this paragraph is in a GroupContentControl.";
            range1.Select();

            groupControl2 = this.Controls.AddGroupContentControl(range1, "groupControl2");
        }
        //</Snippet401>

        // GroupContentControl constructor #3: create GroupContentControl objects for every
        // native group control in the document.
        //<Snippet402>
        private System.Collections.Generic.List
            <Microsoft.Office.Tools.Word.GroupContentControl> groupControls;

        private void CreateGroupControlsFromNativeControls()
        {
            if (this.ContentControls.Count <= 0)
                return;

            groupControls = new System.Collections.Generic.List
                <Microsoft.Office.Tools.Word.GroupContentControl>();
            int count = 0;

            foreach (Word.ContentControl nativeControl in this.ContentControls)
            {
                if (nativeControl.Type == Word.WdContentControlType.wdContentControlGroup)
                {
                    count++;
                    Microsoft.Office.Tools.Word.GroupContentControl tempControl =
                        this.Controls.AddGroupContentControl(nativeControl,
                        "VSTOGroupControl" + count.ToString());
                    groupControls.Add(tempControl);
                }
            }
        }
        //</Snippet402>

        // GroupContentControl constructor #3: create GroupContentControl objects for every
        // native group controls added to the document.
        //<Snippet403>
        void ThisDocument_GroupContentControlAfterAdd(Word.ContentControl NewContentControl, bool InUndoRedo)
        {
            if (NewContentControl.Type == Word.WdContentControlType.wdContentControlGroup)
            {
                this.Controls.AddGroupContentControl(NewContentControl,
                    "GroupControl" + NewContentControl.ID);
            }
        }
        //</Snippet403>

    }
}

File: docs/vsto/codesnippet/VisualBasic/trin_contentcontrolreference/Group.vb
Partial Public NotInheritable Class ThisDocument

    ' GroupContentControl constructor #1: create a GroupContentControl at the current selection.
    '<Snippet400>
    Dim groupControl1 As Microsoft.Office.Tools.Word.GroupContentControl

    Private Sub AddGroupControlAtSelection()
        Me.Paragraphs(1).Range.InsertParagraphBefore()
        Me.Paragraphs(1).Range.Text = "You cannot edit or change the formatting of text " & _
                "in this paragraph, because this paragraph is in a GroupContentControl."
        Me.Paragraphs(1).Range.Select()
        groupControl1 = Me.Controls.AddGroupContentControl("groupControl1")
    End Sub
    '</Snippet400>

    ' GroupContentControl constructor #2: create a GroupContentControl at a specified range.
    '<Snippet401>
    Dim groupControl2 As Microsoft.Office.Tools.Word.GroupContentControl

    Private Sub AddGroupControlAtRange()
        Me.Paragraphs(1).Range.InsertParagraphBefore()
        Dim range1 As Word.Range = Me.Paragraphs(1).Range
        range1.Text = "You cannot edit or change the formatting of text " & _
                "in this paragraph, because this paragraph is in a GroupContentControl."
        range1.Select()
        groupControl2 = Me.Controls.AddGroupContentControl(range1, "groupControl2")
    End Sub
    '</Snippet401>

    ' GroupContentControl constructor #3: create a GroupContentControl for every
    ' native group control in the document.
    '<Snippet402>
    Private groupControls As New System.Collections.Generic.List _
            (Of Microsoft.Office.Tools.Word.GroupContentControl)

    Private Sub CreateGroupControlsFromNativeControls()
        If Me.ContentControls.Count <= 0 Then
            Return
        End If

        Dim count As Integer = 0
        For Each nativeControl As Word.ContentControl In Me.ContentControls
            If nativeControl.Type = Word.WdContentControlType.wdContentControlGroup Then
                count += 1
                Dim tempControl As Microsoft.Office.Tools.Word.GroupContentControl = _
                    Me.Controls.AddGroupContentControl(nativeControl, _
                    "VSTOGroupContentControl" + count.ToString())
                groupControls.Add(tempControl)
            End If
        Next nativeControl
    End Sub
    '</Snippet402>


    ' GroupContentControl constructor #3: create a 
    ' GroupContentControl for every native group added to the document.
    '<Snippet403>
    Private Sub ThisDocument_GroupContentControlAfterAdd(ByVal NewContentControl As Word.ContentControl, _
        ByVal InUndoRedo As Boolean) Handles Me.ContentControlAfterAdd

        If NewContentControl.Type = Word.WdContentControlType.wdContentControlGroup Then
            Me.Controls.AddGroupContentControl(NewContentControl, _
                "GroupControl" + NewContentControl.ID)
        End If
    End Sub
    '</Snippet403>

End Class
